### PR TITLE
adding hide_from_hierachy

### PR DESCRIPTION
**Pull request by pitchandtone**
Very useful if you have a large set of pages and the CMS will only show "Too many pages, show as list". Used on wellingtonnz.com for their listings (accomodation, events, site & activities etc)

### CODE REVIEW

**Comment by pitchandtone** [comment]
Moved from https://github.com/silverstripe/silverstripe-framework/pull/3833

**Comment by micmania1** [comment]
The problem I see with this PR is that front-end functionality would be affected which may not be a desired affect.

*>>begin shameless plug*
https://github.com/micmania1/silverstripe-lumberjack
*>>end shameless plug*

**Comment by wilr** [comment]
See also https://github.com/micschk/silverstripe-excludechildren

**Comment by pitchandtone** [comment]
Above are good outcomes.

**Comment by dhensby** [comment]
I'd say that this probably should still be a core feature. At the moment those modules both still show the currently edited page in the sitetree when editing even if it's meant to be hidden.

So a more complete integration would be better.

**Comment by pitchandtone** [comment]
OK I need to add some tests to it, which is going to be a bit of a bitch.

**Comment by dhensby** [comment]
@pitchandtone In my opinion it also will need an update to the `LeftAndMain` functions that update the site tree (see #3837)

File: model/Hierarchy.php [hunk 1]
```diff
@@ -610,9 +622,13 @@ public function numChildren($cache = true) {
 	 */
 	public function stageChildren($showAll = false) {
 		$baseClass = ClassInfo::baseDataClass($this->owner->class);
+		$hidden = $this->owner->config()->hide_from_hierarchy;
 		$staged = $baseClass::get()
 			->filter('ParentID', (int)$this->owner->ID)
 			->exclude('ID', (int)$this->owner->ID);
+		if (count($hidden) > 0) {
```
Comment (dhensby) [inline model/Hierarchy.php]: `!empty()` is slightly better here as we aren't interested in how many there are, just that there are some

File: model/Hierarchy.php [hunk 2]
```diff
@@ -641,7 +658,9 @@ public function liveChildren($showAll = false, $onlyDeletedFromStage = false) {
 				'Versioned.mode' => $onlyDeletedFromStage ? 'stage_unique' : 'stage',
 				'Versioned.stage' => 'Live'
 			));
-
+		if (count($hidden) > 0) {
```
Comment (dhensby) [inline model/Hierarchy.php]: see https://github.com/silverstripe/silverstripe-framework/pull/3835/files#r24806443

**Comment by dhensby** [comment]
@pitchandtone any progress on this? It'd be a lovely feature IMO

**Comment by pitchandtone** [comment]
What do y'all reckon? Was tossing up adding an extra argument to both stageChildren and LiveChildren which is $includeHidden = false.

**Comment by dhensby** [comment]
There's already "showAll" and I don't think I fully see the need to have an option to show items in the sitetree that have been explicitly hidden for reasons of organisation.

**Comment by pitchandtone** [comment]
Is anyone going to merge this?

File: model/Hierarchy.php [hunk 3]
```diff
@@ -610,9 +622,18 @@ public function numChildren($cache = true) {
 	 */
 	public function stageChildren($showAll = false) {
 		$baseClass = ClassInfo::baseDataClass($this->owner->class);
-		$staged = $baseClass::get()
-			->filter('ParentID', (int)$this->owner->ID)
-			->exclude('ID', (int)$this->owner->ID);
+		$hidden = $this->owner->config()->hide_from_hierarchy;
+		if (empty($hidden)) {
+			$staged = $baseClass::get()
```
Comment (tractorcow) [inline model/Hierarchy.php]:
```php
$staged = $baseClass::get()
	->filter('ParentID', (int)$this->owner->ID)
	->exclude('ID', (int)$this->owner->ID);
if($hidden) {
	$hidden = $hidden->exclude('ClassName', $hidden);
}
```

Also, your double `->exclude(` and `:not` filters negate each other don't they?

File: model/Hierarchy.php [hunk 4]
```diff
@@ -634,15 +655,26 @@ public function liveChildren($showAll = false, $onlyDeletedFromStage = false) {
...
-			));
-
-		if(!$showAll) $children = $children->filter('ShowInMenus', 1);
+		$hidden = $this->owner->config()->hide_from_hierarchy;
+		if (empty($hidden)) {
+			$children = $baseClass::get()
+				->filter('ParentID', (int)$this->owner->ID)
+				->exclude('ID', (int)$this->owner->ID)
+				->setDataQueryParam(array(
+					'Versioned.mode' => $onlyDeletedFromStage ? 'stage_unique' : 'stage',
```
Comment (tractorcow) [inline model/Hierarchy.php]: Same as above; Don't duplicate all of the logic if only a small part of your code changes depending on `$hidden` :)

**Comment by tractorcow** [comment]
I have another question; Don't you think that the `AllChildren` should include filtered children? Looking at the API it feels a bit weird that `Children` is filtered (ok) but `AllChildren` is also filtered? Do we need an `ActuallyAllChildren`? :P

**Comment by pitchandtone** [comment]
I have added an extra static hide_from_cms_tree to cater to both use cases. 

I have also tidied up logic to be more streamlined after @tractorcow's comments.

**Comment by dhensby** [comment]
This needs reading, please

**Comment by tractorcow** [comment]
> I have added an extra static hide_from_cms_tree to cater to both use cases.

Nice, thanks that's all I was worried about really. :)

File: model/Hierarchy.php [hunk 5]
```diff
@@ -38,6 +38,19 @@ class Hierarchy extends DataExtension {
...
+	 * Especially useful for big sets of pages like listings
+	 * If you use this, and still need the classes to be editable
+	 * then add a model admin for the class
+	 * Note: Does not filter subclasses (non-inheriting)
+	 *
+	 * @var array
+	 * @config
+	 */
+	private static $hide_from_hierarchy = array();
+	private static $hide_from_cms_tree = array();
```
Comment (tractorcow) [inline model/Hierarchy.php]: Please split PHPdocs for these. The second var will end up not actually documented.

File: model/Hierarchy.php [hunk 6]
```diff
@@ -602,6 +615,17 @@ public function numChildren($cache = true) {
 	}
 
 	/**
+	 * Checks if we're on a controller where we should filter. ie. Are we loading the SiteTree?
+	 *
+	 * @return bool
+	 */
+	public function showingCMSTree() {
+		$controller = Controller::curr();
```
Comment (tractorcow) [inline model/Hierarchy.php]: `Controller::curr()` throws an error if none is assigned. You can add a `Controller::has_curr()` check to protect against that. :)

**Comment by tractorcow** [comment]
Feedback left. :D

I still feel a bit weird about $allChildren not really being all children, but in most cases it's up to the dev to implement the correct config that's appropriate for their project, and realise the implications, so it's ok.

**Comment by pitchandtone** [comment]
If you're all happy with the above, I'll squash the commits into one.

**Comment by tractorcow** [comment]
Perfect, squash rebase and fix tests please. ;D

**Comment by pitchandtone** [comment]
@tractorcow any idea why we're not getting a result from Travis? I can't see which tests aren't working.

**Comment by pitchandtone** [comment]
Just rebased, that might be it.

**Comment by pitchandtone** [comment]
@tractorcow @dhensby we good to merge?

**Comment by dhensby** [comment]
@pitchandtone it's still saying conflicts!



I'll try to merge by hand

**Comment by dhensby** [comment]
merged by hand

**Comment by micschk** [comment]
I've marked micschk/silverstripe-excludechildren as incompatible from 3.5 upwards & added instructions to the README to use this core functionality instead.

I know this PR is 2 years old by now, but leaving this remark here anyway; I still think the best solution to the 'too many children' problem would be if we could somehow force a page type to use the CMS built-in 'show as list' interface (eg in a 'Children' tab). I don't think this is currently possible so we have to resort to Gridfield/ModelAdmin. But list-view would imo result in a more consistent user experience for this situation.

**Pull request merged**